central/apitoken/expiration/expiration.go

```diff
@@ -0,0 +1,207 @@
+package expiration
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/central/apitoken/datastore"
+	"github.com/stackrox/rox/central/role/resources"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/protoconv"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/search"
+)
+
+const (
+	// The timestamp format / layout is borrowed from `pkg/search/postgres/query/time_query.go`. It would be worth exporting.
+	timestampLayout = "01/02/2006 3:04:05 PM MST"
+)
+
+var (
+	log = logging.LoggerForModule()
+
+	expirySearchCtx = sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+			sac.ResourceScopeKeys(resources.Integration),
+		),
+	)
+
+	scheduleCtx = sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS, storage.Access_READ_WRITE_ACCESS),
+			sac.ResourceScopeKeys(resources.Notifications),
+		),
+	)
+)
+
+// TokenExpirationLoop is the interface for a background task that notifies about API token expiration
+type TokenExpirationLoop interface {
+	Start()
+	Stop()
+}
+
+// TokenExpirationNotifier is the interface for any mechanism that notifies that API Tokens are about to expire
+type TokenExpirationNotifier interface {
+	Notify([]*storage.TokenMetadata) error
+}
+
+type expirationNotifierImpl struct {
+	store datastore.DataStore
+
+	stopper concurrency.Stopper
+
+	notificationTicker *time.Ticker
+
+	notifier TokenExpirationNotifier
+}
+
+func newExpirationNotifier(store datastore.DataStore) *expirationNotifierImpl {
+	return &expirationNotifierImpl{
+		store:    store,
+		stopper:  concurrency.NewStopper(),
+		notifier: &logExpirationNotifier{},
+	}
+}
+
+func (n *expirationNotifierImpl) Start() {
+	if env.PostgresDatastoreEnabled.BooleanSetting() {
+		n.notificationTicker = time.NewTicker(env.APITokenExpirationNotificationInterval.DurationSetting())
+		go n.runExpirationNotifier()
+	}
+}
+
+func (n *expirationNotifierImpl) Stop() {
+	n.stopper.Client().Stop()
+	err := n.stopper.Client().Stopped().Wait()
+	if err != nil {
+		log.Error("Error stopping API Token expiration loop: ", err)
+	}
+}
+
+func (n *expirationNotifierImpl) runExpirationNotifier() {
+	defer n.stopper.Flow().ReportStopped()
+
+	n.checkAndNotifyExpirations()
+
+	if n.notificationTicker == nil {
+		n.notificationTicker = time.NewTicker(env.APITokenExpirationNotificationInterval.DurationSetting())
+	}
```

Comment:
Don't think we need that extra logic here, since we already ensured this within `Start` and this is an unexported method (within the tests, we could just set the ticker "manually" to a relatively low value).